In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
dataset= pd.read_csv("C:\\Users\\pavel\\OneDrive\\Desktop\\fifa21_male2.csv")

C:\Users\pavel\AppData\Local\Temp\ipykernel_11620\431383698.py:1: DtypeWarning: Columns (78) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset= pd.read_csv("C:\\Users\\pavel\\OneDrive\\Desktop\\fifa21_male2.csv")


In [3]:
dataset.head()

ID         Name  Age  OVA            Nationality           Club  BOV   BP  \
0   2  G. Pasquale   33   69                  Italy        Udinese   71  LWB   
1  16  Luis García   37   71                  Spain      KAS Eupen   70   CM   
2  27      J. Cole   33   71                England  Coventry City   71  CAM   
3  36     D. Yorke   36   68  Trinidad &amp; Tobago     Sunderland   70   ST   
4  41      Iniesta   36   81                  Spain    Vissel Kobe   82  CAM   

       Position                                       Player Photo  ...  \
0            LM  https://cdn.sofifa.com/players/000/002/16_120.png  ...   
1    CM CAM CDM  https://cdn.sofifa.com/players/000/016/19_120.png  ...   
2  CAM RM RW LM  https://cdn.sofifa.com/players/000/027/16_120.png  ...   
3           NaN  https://cdn.sofifa.com/players/000/036/09_120.png  ...   
4        CM CAM  https://cdn.sofifa.com/players/000/041/20_120.png  ...   

     CDM    RDM    RWB     LB   LCB    CB   RCB     RB    GK Gender  
0  70+-1  70+-1  71+-2  70+-1  69+0  69+0  69+0  70+-1  17+0   Male  
1   66+1   66+1   62+1   60+1  60+1  60+1  60+1   60+1  17+1   Male  
2   54+0   54+0   52+0   47+0  46+0  46+0  46+0   47+0  15+0   Male  
3   65+0   65+0   56+0   57+0  51+0  51+0  51+0   57+0  22+0   Male  
4   73+3   73+3   70+3   67+3  64+3  64+3  64+3   67+3  17+3   Male  

[5 rows x 107 columns]

In [4]:
dataset.shape

(17125, 107)

In [5]:
pd.set_option('display.max_rows',1000000000 )
pd.set_option('display.max_columns',200)

In [6]:
dataset.isnull().sum()

ID                      0
Name                    0
Age                     0
OVA                     0
Nationality             0
Club                   23
BOV                     0
BP                      0
Position              413
Player Photo            0
Club Logo              23
Flag Photo              0
POT                     0
Team & Contract         0
Height                  0
Weight                  0
foot                    0
Growth                  0
Joined                 58
Loan Date End       16215
Value                   0
Wage                    0
Release Clause          0
Contract                0
Attacking               0
Crossing                0
Finishing               0
Heading Accuracy        0
Short Passing           0
Volleys                58
Skill                   0
Dribbling               0
Curve                  58
FK Accuracy             0
Long Passing            0
Ball Control            0
Movement                0
Acceleration            0
Sprint Speed

#### by looking at the null value,we dont need to handle much of these values as these features will be not be included in our analysis

### lets first drop the unnecessary columns

In [7]:
a= dataset.drop(['BOV','Position','Player Photo','Club Logo','Flag Photo','Team & Contract','Growth','Loan Date End','W/F','SM','A/W','D/W','IR',
'LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB','LB','LCB','CB','RCB','RB','GK','Gender','Hits'
],axis=1)

In [8]:
a.head()

ID         Name  Age  OVA            Nationality           Club   BP  POT  \
0   2  G. Pasquale   33   69                  Italy        Udinese  LWB   69   
1  16  Luis García   37   71                  Spain      KAS Eupen   CM   71   
2  27      J. Cole   33   71                England  Coventry City  CAM   71   
3  36     D. Yorke   36   68  Trinidad &amp; Tobago     Sunderland   ST   82   
4  41      Iniesta   36   81                  Spain    Vissel Kobe  CAM   81   

  Height  Weight   foot        Joined  Value  Wage Release Clause  \
0   6'0"  181lbs   Left   Jul 1, 2008  €625K   €7K             €0   
1  5'10"  143lbs  Right  Jul 19, 2014  €600K   €7K          €1.1M   
2   5'9"  161lbs  Right   Jan 7, 2016  €1.1M  €15K             €0   
3  5'11"  165lbs  Right           NaN     €0    €0             €0   
4   5'7"  150lbs  Right  Jul 16, 2018  €5.5M  €12K          €7.2M   

      Contract  Attacking  Crossing  Finishing  Heading Accuracy  \
0  2008 ~ 2016        313        75         50                59   
1  2014 ~ 2019        337        68         64                61   
2  2016 ~ 2020        337        80         64                41   
3         2009        264        54         70                60   
4  2018 ~ 2021        367        75         69                54   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             71     58.0    338         73   65.0           60            69   
1             76     68.0    369         69   79.0           79            71   
2             77     75.0    387         79   84.0           77            69   
3             80      NaN    255         68    NaN           46            64   
4             90     79.0    408         85   80.0           70            83   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            71       347            68            74     68.0         69   
1            71       305            56            50     62.0         65   
2            78       295            48            42     71.0         59   
3            77       176            59            62      NaN         55   
4            90       346            61            56     79.0         75   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0     68.0    347          74     68.0       69        68          68   
1     72.0    324          75     54.0       64        60          71   
2     75.0    284          72     58.0       29        56          69   
3      NaN    239          63      NaN       51        66          59   
4     75.0    297          67     40.0       58        62          70   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        320          72           69.0         63.0    66.0         50   
1        362          71           71.0         72.0    73.0         75   
2        317          69           39.0         69.0    74.0         66   
3        271          59           70.0         72.0     NaN         70   
4        370          58           70.0         78.0    93.0         71   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0        NaN        208       70               69            69.0   
1       79.0        153       70               43            40.0   
2        NaN         99       35               34            30.0   
3        NaN         75       34               41             NaN   
4       89.0        181       68               57            56.0   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           56         14            5          15              10   
1           56          9           12          13              11   
2           51          9            6          13              16   
3           68          5           21          64              21   
4           45          6           13           6              13   

   GK Reflexes  Total 

In [9]:
a.shape

(17125, 65)

In [10]:
a.dtypes

ID                    int64
Name                 object
Age                   int64
OVA                   int64
Nationality          object
Club                 object
BP                   object
POT                   int64
Height               object
Weight               object
foot                 object
Joined               object
Value                object
Wage                 object
Release Clause       object
Contract             object
Attacking             int64
Crossing              int64
Finishing             int64
Heading Accuracy      int64
Short Passing         int64
Volleys             float64
Skill                 int64
Dribbling             int64
Curve               float64
FK Accuracy           int64
Long Passing          int64
Ball Control          int64
Movement              int64
Acceleration          int64
Sprint Speed          int64
Agility             float64
Reactions             int64
Balance             float64
Power                 int64
Shot Power          

### Now lets handle datatypes some of the features

#### Height 

In [11]:
def height_conversor(Height):
    feet, inches = Height.split("'")
    feet = int(feet)
    inches = int(inches.strip('"'))
    total_inches = feet * 12 + inches
    cm = total_inches * 2.54
    return cm
a['Height_cm']=a['Height'].apply(height_conversor)


## Weight

In [12]:
def weight_conversor(Weight):
    weight_kgs=float(Weight.split('lbs')[0].strip())
    return weight_kgs * 0.45359237     

In [13]:
a['Weight_kgs']=a['Weight'].apply(weight_conversor)

### Value,Wage,Release Clause

In [14]:
def clean_cols(value):
    value=value.lstrip('€')
    suffix=value[-1]
    if suffix=='M':
        return float(value[:-1])*1000000
    elif suffix=='K':
        return float (value[:-1])*1000
    else:
        return float(value)

In [15]:
a['Values_in_euros']=a['Value'].apply(clean_cols)
a['Wage_in_euros']=a['Wage'].apply(clean_cols)
a['Release_clause_euros']=a['Release Clause'].apply(clean_cols)

In [16]:
df=a.drop(['Height','Weight','Value','Wage','Release Clause'],axis=1)

In [17]:
df.head()

ID         Name  Age  OVA            Nationality           Club   BP  POT  \
0   2  G. Pasquale   33   69                  Italy        Udinese  LWB   69   
1  16  Luis García   37   71                  Spain      KAS Eupen   CM   71   
2  27      J. Cole   33   71                England  Coventry City  CAM   71   
3  36     D. Yorke   36   68  Trinidad &amp; Tobago     Sunderland   ST   82   
4  41      Iniesta   36   81                  Spain    Vissel Kobe  CAM   81   

    foot        Joined     Contract  Attacking  Crossing  Finishing  \
0   Left   Jul 1, 2008  2008 ~ 2016        313        75         50   
1  Right  Jul 19, 2014  2014 ~ 2019        337        68         64   
2  Right   Jan 7, 2016  2016 ~ 2020        337        80         64   
3  Right           NaN         2009        264        54         70   
4  Right  Jul 16, 2018  2018 ~ 2021        367        75         69   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                59             71     58.0    338         73   65.0   
1                61             76     68.0    369         69   79.0   
2                41             77     75.0    387         79   84.0   
3                60             80      NaN    255         68    NaN   
4                54             90     79.0    408         85   80.0   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           60            69            71       347            68   
1           79            71            71       305            56   
2           77            69            78       295            48   
3           46            64            77       176            59   
4           70            83            90       346            61   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            74     68.0         69     68.0    347          74     68.0   
1            50     62.0         65     72.0    324          75     54.0   
2            42     71.0         59     75.0    284          72     58.0   
3            62      NaN         55      NaN    239          63      NaN   
4            56     79.0         75     75.0    297          67     40.0   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       69        68          68        320          72           69.0   
1       64        60          71        362          71           71.0   
2       29        56          69        317          69           39.0   
3       51        66          59        271          59           70.0   
4       58        62          70        370          58           70.0   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0         63.0    66.0         50        NaN        208       70   
1         72.0    73.0         75       79.0        153       70   
2         69.0    74.0         66        NaN         99       35   
3         72.0     NaN         70        NaN         75       34   
4         78.0    93.0         71       89.0        181       68   

   Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
0               69            69.0           56         14            5   
1               43            40.0           56          9           12   
2               34            30.0           51          9            6   
3               41             NaN           68          5           21   
4               57            56.0           45          6           13   

   GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  PAC  SHO  \
0          15              10           12         1929         408   71   59   
1          13              11           11         1906         385   53   69   
2          13              16            7         1770         354   45   68   
3          64              21           21         1348         369   61   66   
4           6              13            7         2014         420   58  

In [18]:
df=df.drop(['Joined'],axis=1)

# Splitting Contract Column

In [19]:
df[['joined_year','Leaving_year']]=df['Contract'].str.split('~',expand=True)


In [20]:
df['joined_year']=pd.to_numeric(df['joined_year'],errors='coerce')
df['Leaving_year']=pd.to_numeric(df['Leaving_year'],errors='coerce')

In [21]:
df=df.drop(['Contract'],axis=1)


In [22]:
df.head()

ID         Name  Age  OVA            Nationality           Club   BP  POT  \
0   2  G. Pasquale   33   69                  Italy        Udinese  LWB   69   
1  16  Luis García   37   71                  Spain      KAS Eupen   CM   71   
2  27      J. Cole   33   71                England  Coventry City  CAM   71   
3  36     D. Yorke   36   68  Trinidad &amp; Tobago     Sunderland   ST   82   
4  41      Iniesta   36   81                  Spain    Vissel Kobe  CAM   81   

    foot  Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  \
0   Left        313        75         50                59             71   
1  Right        337        68         64                61             76   
2  Right        337        80         64                41             77   
3  Right        264        54         70                60             80   
4  Right        367        75         69                54             90   

   Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  \
0     58.0    338         73   65.0           60            69            71   
1     68.0    369         69   79.0           79            71            71   
2     75.0    387         79   84.0           77            69            78   
3      NaN    255         68    NaN           46            64            77   
4     79.0    408         85   80.0           70            83            90   

   Movement  Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  \
0       347            68            74     68.0         69     68.0    347   
1       305            56            50     62.0         65     72.0    324   
2       295            48            42     71.0         59     75.0    284   
3       176            59            62      NaN         55      NaN    239   
4       346            61            56     79.0         75     75.0    297   

   Shot Power  Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  \
0          74     68.0       69        68          68        320          72   
1          75     54.0       64        60          71        362          71   
2          72     58.0       29        56          69        317          69   
3          63      NaN       51        66          59        271          59   
4          67     40.0       58        62          70        370          58   

   Interceptions  Positioning  Vision  Penalties  Composure  Defending  \
0           69.0         63.0    66.0         50        NaN        208   
1           71.0         72.0    73.0         75       79.0        153   
2           39.0         69.0    74.0         66        NaN         99   
3           70.0         72.0     NaN         70        NaN         75   
4           70.0         78.0    93.0         71       89.0        181   

   Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  \
0       70               69            69.0           56         14   
1       70               43            40.0           56          9   
2       35               34            30.0           51          9   
3       34               41             NaN           68          5   
4       68               57            56.0           45          6   

   GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  \
0            5          15              10           12         1929   
1           12          13              11           11         1906   
2            6          13              16            7         1770   
3           21          64              21           21         1348   
4           13           6              13            7         2014   

   Base Stats  PAC  SHO  PAS  DRI  DEF  PHY  Height_cm  Weight_kgs  \
0         408   71   59   70   71   68   69     182.88   82.100219   
1         385   53   69   73   69   58   63     177.80   64.863709   
2         354   45   68   76   77   36   52     175.26   73.028372   
3         369   61   66   66   69   4

In [23]:
df=df.drop(['Composure','joined_year','Leaving_year'],axis=1)


# Regression model to predict rating of players
###### As we are taking a lot of input variables we will be using Lasso Regeression as it bring most of the coefficient to zero and only takes important variables

In [24]:
from sklearn.model_selection import train_test_split

In [26]:
#defining independent variables
x= df.iloc[:,9:59]

In [27]:
for col in x.columns:
    if x[col].isnull().any():
        x[col].fillna(x[col].mean(),inplace=True)
    print(x.head())

   Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  \
0        313        75         50                59             71     58.0   
1        337        68         64                61             76     68.0   
2        337        80         64                41             77     75.0   
3        264        54         70                60             80      NaN   
4        367        75         69                54             90     79.0   

   Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
0    338         73   65.0           60            69            71       347   
1    369         69   79.0           79            71            71       305   
2    387         79   84.0           77            69            78       295   
3    255         68    NaN           46            64            77       176   
4    408         85   80.0           70            83            90       346   

   Acceleration  Sprint Speed  Agility

4         420   58   70   85   85   63   59     170.18   68.038855  
   Attacking  Crossing  Finishing  Heading Accuracy  Short Passing    Volleys  \
0        313        75         50                59             71  58.000000   
1        337        68         64                61             76  68.000000   
2        337        80         64                41             77  75.000000   
3        264        54         70                60             80  45.005742   
4        367        75         69                54             90  79.000000   

   Skill  Dribbling      Curve  FK Accuracy  Long Passing  Ball Control  \
0    338         73  65.000000           60            69            71   
1    369         69  79.000000           79            71            71   
2    387         79  84.000000           77            69            78   
3    255         68  49.574676           46            64            77   
4    408         85  80.000000           70            83            

In [28]:
x.isnull().sum()

Attacking           0
Crossing            0
Finishing           0
Heading Accuracy    0
Short Passing       0
Volleys             0
Skill               0
Dribbling           0
Curve               0
FK Accuracy         0
Long Passing        0
Ball Control        0
Movement            0
Acceleration        0
Sprint Speed        0
Agility             0
Reactions           0
Balance             0
Power               0
Shot Power          0
Jumping             0
Stamina             0
Strength            0
Long Shots          0
Mentality           0
Aggression          0
Interceptions       0
Positioning         0
Vision              0
Penalties           0
Defending           0
Marking             0
Standing Tackle     0
Sliding Tackle      0
Goalkeeping         0
GK Diving           0
GK Handling         0
GK Kicking          0
GK Positioning      0
GK Reflexes         0
Total Stats         0
Base Stats          0
PAC                 0
SHO                 0
PAS                 0
DRI       

In [29]:
y=df.loc[:,['OVA']]

In [30]:
x.head()

Attacking  Crossing  Finishing  Heading Accuracy  Short Passing    Volleys  \
0        313        75         50                59             71  58.000000   
1        337        68         64                61             76  68.000000   
2        337        80         64                41             77  75.000000   
3        264        54         70                60             80  45.005742   
4        367        75         69                54             90  79.000000   

   Skill  Dribbling      Curve  FK Accuracy  Long Passing  Ball Control  \
0    338         73  65.000000           60            69            71   
1    369         69  79.000000           79            71            71   
2    387         79  84.000000           77            69            78   
3    255         68  49.574676           46            64            77   
4    408         85  80.000000           70            83            90   

   Movement  Acceleration  Sprint Speed    Agility  Reactions    Balance  \
0       347            68            74  68.000000         69  68.000000   
1       305            56            50  62.000000         65  72.000000   
2       295            48            42  71.000000         59  75.000000   
3       176            59            62  64.602273         55  64.715591   
4       346            61            56  79.000000         75  75.000000   

   Power  Shot Power    Jumping  Stamina  Strength  Long Shots  Mentality  \
0    347          74  68.000000       69        68          68        320   
1    324          75  54.000000       64        60          71        362   
2    284          72  58.000000       29        56          69        317   
3    239          63  65.170212       51        66          59        271   
4    297          67  40.000000       58        62          70        370   

   Aggression  Interceptions  Positioning     Vision  Penalties  Defending  \
0          72           69.0         63.0  66.000000         50        208   
1          71           71.0         72.0  73.000000         75        153   
2          69           39.0         69.0  74.000000         66         99   
3          59           70.0         72.0  55.443605         70         75   
4          58           70.0         78.0  93.000000         71        181   

   Marking  Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  \
0       70               69       69.000000           56         14   
1       70               43       40.000000           56          9   
2       35               34       30.000000           51          9   
3       34               41       46.099725           68          5   
4       68               57       56.000000           45          6   

   GK Handling  GK Kicking  GK Positioning  GK Reflexes  Total Stats  \
0            5          15              10           12         1929   
1           12          13              11           11         1906   
2            6          13              16            7         1770   
3           21          64              21           21         1348   
4           13           6              13            7         2014   

   Base Stats  PAC  SHO  PAS  DRI  DEF  PHY  Height_cm  Weight_kgs  
0         408   71   59   70   71   68   69     182.88   82.100219  
1         385   53   69   73   69   58   63     177.80   64.863709  
2         354   45   68   76   77   36   52     175.26   73.028372  
3         369   61   66   66   69   47   60     180.34   74.842741  
4         420   58   70   85   85   63   59     170.18   68.038855

##### Dropping Height and Weight as it does not affect a players skills much

In [31]:
x=x.drop(['Height_cm','Weight_kgs'],axis=1)

In [32]:
x.shape

(17125, 48)

In [33]:
y.head()

OVA
0   69
1   71
2   71
3   68
4   81

In [34]:
y.value_counts().sum()

17125

In [35]:
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split

In [36]:
 X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

## Standardizing the Scale


In [39]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [40]:
Las = Lasso(alpha=0.01)

In [41]:
 Las.fit(X_train_scaled,y_train)

Lasso(alpha=0.01)

In [42]:
prediction= Las.predict(X_test_scaled)

In [43]:
from sklearn.metrics import r2_score

In [44]:
r2_score(y_test,prediction)

0.8815188086985484

In [45]:
from sklearn.model_selection import cross_val_score

In [46]:
print("Accuracy",np.mean(cross_val_score(Las,x,y,cv=5,scoring="r2")))

C:\Users\pavel\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.563e+04, tolerance: 6.123e+01
  model = cd_fast.enet_coordinate_descent(
C:\Users\pavel\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.268e+04, tolerance: 6.580e+01
  model = cd_fast.enet_coordinate_descent(
C:\Users\pavel\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.931e+04, toleranc

Accuracy 0.8152647868377295


C:\Users\pavel\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.907e+04, tolerance: 5.368e+01
  model = cd_fast.enet_coordinate_descent(


In [79]:
# Print non-zero coefficients (selected features)
selected_features = x.columns[Las.coef_ != 0]
print('Selected Features:', selected_features)

Selected Features: Index(['Crossing', 'Heading Accuracy', 'Short Passing', 'Volleys', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Acceleration',
       'Sprint Speed', 'Agility', 'Reactions', 'Balance', 'Shot Power',
       'Jumping', 'Stamina', 'Strength', 'Long Shots', 'Mentality',
       'Aggression', 'Positioning', 'Vision', 'Penalties', 'Marking',
       'Sliding Tackle', 'GK Handling', 'GK Positioning', 'Base Stats', 'PAC',
       'SHO', 'PAS', 'DRI', 'PHY', 'Height_cm', 'Weight_kgs'],
      dtype='object')


# We got and accuracy of 88% in pridicting the Overall Rating of  Players 